crates/rust-analyzer/src/diff.rs

```diff
@@ -0,0 +1,61 @@
+use dissimilar::Chunk;
+use ide::{TextEdit, TextRange, TextSize};
+
+fn textedit_from_chunks(chunks: Vec<dissimilar::Chunk>) -> TextEdit {
+    let mut builder = TextEdit::builder();
+    let mut pos = TextSize::default();
+
+    let mut pending_delete_start = None;
+
+    for chunk in chunks {
+        match (pending_delete_start, chunk) {
+            (Some(delete_start), Chunk::Insert(text)) => {
+                builder.replace(TextRange::new(delete_start, pos), text.into());
+                pending_delete_start = None;
+            }
+            (_, _) => {
+                if let Some(delete_start) = pending_delete_start {
+                    builder.delete(TextRange::new(delete_start, pos));
+                    pending_delete_start = None;
+                }
+
+                match chunk {
+                    Chunk::Equal(text) => {
+                        pos += TextSize::of(text);
+                    }
+                    Chunk::Delete(text) => {
+                        pending_delete_start = Some(pos);
+                        pos += TextSize::of(text);
+                    }
+                    Chunk::Insert(text) => {
+                        builder.insert(pos, text.into());
+                        pos += TextSize::of(text);
+                    }
+                }
+            }
+        }
+    }
```

Comment:
Would using  `chunks.peekable` help here? Seems like we could detect delet-insert pairs that way.

Reply:
I see two ways how that could work. One is to move ownership of the iterator out of the loop and skip the the insert iteration if it is peeked, but that gets pretty gnarly. I tried that, and the general logic gets a bit simpler, but the control flow gets difficult to follow. The other way would be to set a variable that an iteration should be skipped, which would look very similar to the current implementation. 

A solution that's maybe clearer would be to introduce an enum similar to Chunk, but with Replace, then Iterator::map chunks to that and Itertools::coalesce the result and map that to Indel

Reply:
Hm, 

```rust
fn textedit_from_chunks(chunks: Vec<dissimilar::Chunk>) -> TextEdit {
    let mut builder = TextEdit::builder();
    let mut pos = TextSize::default();

    let mut chunks = chunks.into_iter().peekable();
    while let Some(ch) = chunks.next() {
        if let (Chunk::Delete(del), Some(&Chunk::Insert(ins))) = (ch, chunks.peek()) {
            chunks.next().unwrap();
            let del_len = TextSize::of(del);
            builder.replace(TextRange::at(pos, del_len), ins.into());
            pos += del_len;
            continue;
        }

        match ch {
            Chunk::Equal(text) => {
                pos += TextSize::of(text);
            }
            Chunk::Delete(del) => {
                let del_len = TextSize::of(del);
                builder.delete(TextRange::at(pos, del_len));
                pos += del_len;
            }
            Chunk::Insert(text) => {
                builder.insert(pos, text.into());
            }
        }
    }

    builder.finish()
}
```

seems a bit simpler to me....

Reply:
Yes, you are right, that is simpler. I probably overcomplicated it a bit when I tried it that way :)